java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java

```diff
@@ -0,0 +1,1402 @@
+   * case where index and length of a new element are beyond the existing
+   * capacity of the vector.
+   *
+   * @param index   position of the element to set
+   * @param value   array of bytes to write
+   * @param start   start index in an array of bytes
+   * @param length  length of data in an array of bytes
+   */
+  public void setSafe(int index, byte[] value, int start, int length) {
+    assert index >= 0;
+    handleSafe(index, length);
+    BitVectorHelper.setBit(validityBuffer, index);
+    setBytes(index, value, start, length);
+    lastSet = index;
+  }
+
+  /**
+   * Set the variable length element at the specified index to the
+   * content in supplied ByteBuffer.
+   *
+   * @param index   position of the element to set
+   * @param value   ByteBuffer with data
+   * @param start   start index in ByteBuffer
+   * @param length  length of data in ByteBuffer
+   */
+  public void set(int index, ByteBuffer value, int start, int length) {
+    assert index >= 0;
+    BitVectorHelper.setBit(validityBuffer, index);
+    setBytes(index, value.array(), start, length);
+    lastSet = index;
+  }
+
+  /**
+   * Same as {@link #set(int, ByteBuffer, int, int)} except that it handles the
+   * case where index and length of a new element are beyond the existing
+   * capacity of the vector.
+   *
+   * @param index   position of the element to set
+   * @param value   ByteBuffer with data
+   * @param start   start index in ByteBuffer
+   * @param length  length of data in ByteBuffer
+   */
+  public void setSafe(int index, ByteBuffer value, int start, int length) {
+    assert index >= 0;
+    handleSafe(index, length);
+    BitVectorHelper.setBit(validityBuffer, index);
+    setBytes(index, value.array(), start, length);
+    lastSet = index;
+  }
+
+  /**
+   * Set the element at the given index to null.
+   *
+   * @param index position of an element
+   */
+  @Override
+  public void setNull(int index) {
+    // We need to check and reallocate the validity buffer
+    while (index >= getValueCapacity()) {
+      reallocValidityBuffer();
+    }
+    BitVectorHelper.unsetBit(validityBuffer, index);
+  }
+
+  /**
+   * Store the given value at a particular position in the vector. isSet indicates
+   * whether the value is NULL or not.
+   * @param index position of the new value
+   * @param isSet Zero for NULL value, 1 otherwise
+   * @param start start position of data in buffer
+   * @param end end position of data in buffer
+   * @param buffer data buffer containing the variable width element to be stored
+   *               in the vector
+   */
+  public void set(int index, int isSet, int start, int end, ArrowBuf buffer) {
+    assert index >= 0;
+    final int dataLength = end - start;
+    BitVectorHelper.setValidityBit(validityBuffer, index, isSet);
+    byte[] data = new byte[dataLength];
+    buffer.getBytes(start, data, 0, dataLength);
+    setBytes(index, data, start, dataLength);
```

Comment:
We should be able to avoid this copy?

Reply:
We can introduce an overload like `protected final void setBytes(int index, ArrowBuf valueBuf, int start, int length)` and avoid that additional copy.

Reply:
There is a degree of code duplicates in the updated version. Although I tried to abstract it away as follows, but I got an exception.

Failed abstraction attempt 

```java
protected final void setBytes(int index, byte[] value, int start, int length) {
    ArrowBuf currentBuf = null;
    if (length > INLINE_SIZE) {
        currentBuf = allocateOrGetLastDataBuffer(length);
        currentBuf.setBytes(currentBuf.writerIndex(), value, start, length);
        currentBuf.writerIndex(currentBuf.writerIndex() + length);
    }
    setViewBuffer(index, value, start, length, currentBuf);
}

protected final void setBytes(int index, ArrowBuf valueBuf, int start, int length) {
    ArrowBuf currentBuf = null;
    if (length > INLINE_SIZE) {
        currentBuf = allocateOrGetLastDataBuffer(length);
        currentBuf.setBytes(currentBuf.writerIndex(), valueBuf, start, length);
        currentBuf.writerIndex(currentBuf.writerIndex() + length);
    }
    setViewBuffer(index, valueBuf.nioBuffer().array(), start, length, currentBuf);
}

private void setViewBuffer(int index, byte[] value, int start, int length, ArrowBuf currentBuf) {
    int writePosition = index * ELEMENT_SIZE;
    viewBuffer.setInt(writePosition, length);
    writePosition += LENGTH_WIDTH;
    if (length <= INLINE_SIZE) {
        viewBuffer.setBytes(writePosition, value, start, length);
    } else {
        viewBuffer.setBytes(writePosition, value, start, PREFIX_WIDTH);
        writePosition += PREFIX_WIDTH;
        viewBuffer.setInt(writePosition, dataBuffers.size() - 1);
        writePosition += BUF_INDEX_WIDTH;
        viewBuffer.setInt(writePosition, (int) currentBuf.writerIndex());
    }
}
```

```bash
java.lang.UnsupportedOperationException
	at java.base/java.nio.ByteBuffer.array(ByteBuffer.java:1505)
at org.apache.arrow.vector/org.apache.arrow.vector.BaseVariableWidthViewVector.setBytes(BaseVariableWidthViewVector.java:1245)
	at org.apache.arrow.vector/    org.apache.arrow.vector.BaseVariableWidthViewVector.setSafe(BaseVariableWidthViewVector.java:1115)
	at org.apache.arrow.vector/org.apache.arrow.vector.TestValueVector.testSetSafeWithArrowBufNoExcessAllocs(TestValueVector.java:1978)

```

failed at calling -> `valueBuf.nioBuffer().array()`

I wasn't sure, thus I settled for a bit code duplicated approach where `setBytes` has 2 versions which supports `byte[]` input and `ArrowBuf` as an input. Can I do better here?

Reply:
I don't think it matters that much to duplicate a ~dozen lines of code